src/blockencodings.cpp

```diff
@@ -15,21 +16,45 @@
 
 #include <unordered_map>
 
+uint64_t GetShortID(
+        const uint64_t& shorttxidk0,
+        const uint64_t& shorttxidk1,
+        const uint256& txhash)
+{
+    static_assert(CompactBlock::SHORTTXIDS_LENGTH == 6, "shorttxids calculation assumes 6-byte shorttxids");
+    return SipHashUint256(shorttxidk0, shorttxidk1, txhash) & 0xffffffffffffL;
+}
+
 #define MIN_TRANSACTION_SIZE (::GetSerializeSize(CTransaction(), SER_NETWORK, PROTOCOL_VERSION))
 
-CBlockHeaderAndShortTxIDs::CBlockHeaderAndShortTxIDs(const CBlock& block) :
-        nonce(GetRand(std::numeric_limits<uint64_t>::max())),
-        shorttxids(block.vtx.size() - 1), prefilledtxn(1), header(block) {
+CompactBlock::CompactBlock(const CBlock& block,
+    const CRollingBloomFilter* inventoryKnown) :
+        nonce(GetRand(std::numeric_limits<uint64_t>::max())), header(block)
+{
     FillShortTxIDSelector();
-    //TODO: Use our mempool prior to block acceptance to predictively fill more than just the coinbase
-    prefilledtxn[0] = {0, block.vtx[0]};
+
+    if (block.vtx.empty())
+        throw std::invalid_argument(__func__ + std::string(" expects coinbase tx"));
+
+    //< Index of a prefilled tx is its diff from last index.
+    size_t prevIndex = 0;
+    prefilledtxn.push_back(PrefilledTransaction{0, block.vtx[0]});
+
     for (size_t i = 1; i < block.vtx.size(); i++) {
         const CTransaction& tx = block.vtx[i];
-        shorttxids[i - 1] = GetShortID(tx.GetHash());
+
+        if (inventoryKnown && !inventoryKnown->contains(tx.GetHash()))
```

Comment:
This at a minimum this should limit the amount it sends, and not use this heuristic when the peer has transaction relay disabled (!fRelayTxes).

BIP152 states:
"Nodes sending cmpctblock messages SHOULD limit prefilledtxn to 10KB of transactions. When in doubt, nodes SHOULD only include the coinbase transaction in prefilledtxn."

If it's going to send more than that, the latency is already hosed, and it might as well take the extra round trip and not waste bandwidth.

Filterinventoryknown isn't going to match any transactions if the peer is blocks only, so as implemented this will spam them with the whole block, and waste an epic ton of bandwidth for no good reason.

What we tested while defining the protocol was using the list of transactions which were misses on this node when it received the block. This was highly effective and had a low false positive rate, you might want to consider that. You could further augment it with leaving out things the far end has advertised-- we never tried that, but it would be an interesting experiment.

In general, I found the HB mode to be much more effective at reducing latency, with a somewhat lower bandwidth penalty than extensive prediction; but I'm glad you're experimenting here.